tests/client-dbus/tests/dbus/pool/test_udev.py

```diff
@@ -0,0 +1,411 @@
+
+    def test_existing(self):
+        """
+        While we are adding devices back we will stop start the daemon to ensure
+        it can start with one or more devices present and complete when the
+        other devices come in later.
+        :return: None
+        """
+        self._test_driver(2, 4, True)
+
+    def _single_pool(self, num_devices, num_hotplugs=0):
+        """
+        Creates a single pool with specified number of devices.
+        :param num_devices: Number of devices to use for pool
+        :param num_hotplugs: Number of extra udev "add" event per devices
+        :return: None
+        """
+        self._start_service()
+        result = UdevAdd._get_pools()
+        self.assertEqual(len(result), 0)
+
+        device_tokens = []
+
+        for _ in range(0, num_devices):
+            device_tokens.append(self._lb_mgr.create_device())
+
+        self.assertEqual(len(device_tokens), num_devices)
+
+        pool_name = rs(5)
+        UdevAdd._create_pool(pool_name, self._device_files(device_tokens))
+
+        self.assertEqual(len(UdevAdd._get_pools()), 1)
+
+        self._stop_service_remove_dm_tables()
+        self._start_service()
+
+        # Make sure on a start with all the devices the pool is there!
+        self.assertEqual(len(UdevAdd._get_pools()), 1)
+
+        self._stop_service_remove_dm_tables()
+
+        # Remove the devices
+        for d in device_tokens:
+            self._lb_mgr.unplug(d)
+
+        self._start_service()
+
+        self.assertEqual(len(UdevAdd._get_pools()), 0)
+
+        for d in device_tokens:
+            self._lb_mgr.hotplug(d)
+
+        self._settle()
+
+        self.assertEqual(len(UdevAdd._get_pools()), 1)
+
+        # Generate unnecessary hot plug adds
+        for _ in range(0, num_hotplugs):
+            for d in device_tokens:
+                self._lb_mgr.generate_udev_add_event(d)
+
+        self._settle()
+
+        self.assertEqual(len(UdevAdd._get_pools()), 1)
+
+    def test_simultaneous(self):
+        """
+        Create a single pool with 16 devices and simulate them being hotplug
+        at same time
+        :return: None
+        """
+        self._single_pool(16)
+
+    def test_spurious_adds(self):
+        """
+        Create a single pool with 16 devices and simulate them being hotplug
+        at same time and with spurious additional "add" udev events
+        :return: None
+        """
+        self._single_pool(16, 4)
+
+    def test_simple_udev_add(self):
+        """
+        Create a single pool with 1 device!
+        :return: None
+        """
+        self._single_pool(1, 1)
+
+    def test_duplicate_pool_name(self):
+        """
+        Create more than one pool with the same name, then dynamically fix it
+        :return: None
+        """
+        pool_name = rs(12)
+        pool_tokens = []
+        num_pools = 3
+
+        self._start_service()
+
+        # Create some pools with duplicate names
```

Comment:
I'm stopping the service, removing all the dm tables and detaching the block devices, then restarting after each pool create.  Thus each pool that gets created has no knowledge of each other.  This was done so that I can start the service and systematically hot-plug add the disks to force the duplicate pool names found.

Reply:
Aaah!